kinds.go

```diff
@@ -0,0 +1,118 @@
+package kindsys
+
+import (
+	"context"
+	"io"
+)
+
+type KindInfo struct {
+	// Organization controlled prefix
+	Group string `json:"group"`
+
+	// Must be unique within the group
+	Kind string `json:"kind"`
+
+	// Description of the purpose of this kind
+	Description string `json:"description"`
+
+	// This kind depends on composable types
+	UsesComposition []string `json:"usesComposition,omitempty"`
+
+	// Indicate where this kind is in the dev cycle
+	Maturity Maturity `json:"maturity"`
+}
+
+type VersionInfo struct {
+	// Must be vMajor-Minor-alpha
+	// for k8s... we can't have "." and support aggregation api
+	Version string `json:"version"`
+
+	// The software version when this schema was released.
+	// NOTE(1): the version must follow semantic versioning so that the order is deterministic
+	// NOTE(2): panel plugin version is saved in dashboards.  This can be used
+	// to find the appropriate schema
+	SoftwareVersion string `json:"software"`
+
+	// Human readable descriptions of the changes in this version
+	Changelog []string `json:"changelog,omitempty"`
+
+	// The YYYY-MM-DD this version was published (or empty if not yet published)
+	Published string `json:"published,omitempty"`
+
+	// JSONSchema hash
+	Signature string `json:"signature,omitempty"` // ?? hash of the json schema
+}
+
+type MachineNames struct {
+	// This is used in k8s URLs
+	Plural string `json:"plural,omitempty"`
+
+	// Used as an alias in the display
+	Singular string `json:"singular,omitempty"`
+
+	// Optional shorter names that can be matched in a CLI
+	Short []string `json:"short,omitempty"`
+}
+
+type Kind interface {
+	// Get general information about this kind
+	GetKindInfo() KindInfo
+
+	// Get the latest version
+	CurrentVersion() string
+
+	// Get all versions
+	GetVersions() []VersionInfo
+
+	// Return a JSON schema definition for the selected version
+	// When composition is required, the slots will have an any node
+	// TODO? include an option to have `AnyOf(latest known options)`
+	GetJSONSchema(version string) (string, error)
+}
+
+// A kind that manages a k8s style resource
+type ResourceKind interface {
+	Kind
+
+	// K8S style machine names for this kind
+	GetMachineNames() MachineNames
+
+	// Read data into a Resource, when strict is true, all validation rules will be checked
+	// otherwise a resource will be created if possible, but all validation may not have been run
+	Read(reader io.Reader, strict bool) (Resource, error)
+
+	// Migrate from one object to another version
+	// NOTE, this may require database calls
+	Migrate(ctx context.Context, obj Resource, targetVersion string) (Resource, error)
+}
+
+// Function that can be used to migrate resources
+type ResourceMigrator = func(ctx context.Context, obj Resource, targetVersion string) (Resource, error)
+
+type ComposableKind interface {
+	Kind
+
+	// eg: panel(options+fieldconfig) | transformation | dataquery | matcher
+	GetComposableType() string
```

Comment:
What is `ComposableType` in this context? couldn't we just use `kind.name`?

Reply:
This still needs exploration -- and I think should be an explicit enumeration.  This is essentially what the kind implements.

eg:
ComposableType = PanelOptions
Kind.name = timeseries | heatmap | stats | ...

Then we can ask for all ComposableKinds of type "PanelOptions" when constructing/validating panels

Reply:
currently (before this PR) each "ComposableKind" has an [schemaInterface](https://github.com/grafana/kindsys/blob/74f2bc8126104cd5c989a8d0e2a8539f6a8a8b2e/props.go#L68C48-L68C63) -- that is the same as type here, and "slot" is used to say a sub-element.

The only place this is used is a "panel" schemaInterface that has two slots: options and fieldConfig.  I think it is worth exploring if these can just be two kinds (PanelOptions and PanelFieldConfig?) rather than having both type/schemaInterface *and* slots concept.    but again... this will take some effort to see how using it for real things feels before we should really decide